datreant/backends/statefiles.py

```diff
@@ -0,0 +1,448 @@
+                user-given name of Treant object
+            *categories*
+                user-given dictionary with custom keys and values; used to
+                give distinguishing characteristics to object for search
+            *tags*
+                user-given list with custom elements; used to give
+                distinguishing characteristics to object for search
+            *version*
+                version of datreant file was generated with
+        """
+        # update schema and version of file
+        version = self.update_schema()
+        self.update_version(version)
+
+        # tags table
+        tags = kwargs.pop('tags', list())
+        self.add_tags(*tags)
+
+        # categories table
+        categories = kwargs.pop('categories', dict())
+        self.add_categories(**categories)
+
+    @FileSerial._read
+    @FileSerial._pull
+    def get_version(self):
+        """Get Treant version.
+
+        :Returns:
+            *version*
+                version of Treant
+
+        """
+        return self._record['version']
+
+    # TODO: need a proper schema update mechanism
+    @FileSerial._write
+    @FileSerial._pull_push
+    def update_schema(self):
+        """Update schema of file.
+
+        :Returns:
+            *version*
+                version number of file's new schema
+        """
+        try:
+            version = self._record['version']
+        except KeyError:
+            version = datreant.__version__
+
+        return version
+
+    @FileSerial._write
+    @FileSerial._pull_push
+    def update_version(self, version):
+        """Update version of Treant.
+
+        :Arugments:
+            *version*
+                new version of Treant
+        """
+        self._record['version'] = version
+
+    @FileSerial._read
+    @FileSerial._pull
+    def get_tags(self):
+        """Get all tags as a list.
+
+        :Returns:
+            *tags*
+                list of all tags
+        """
+        return self._record['tags']
+
+    @FileSerial._write
+    @FileSerial._pull_push
+    def add_tags(self, *tags):
+        """Add any number of tags to the Treant.
+
+        Tags are individual strings that serve to differentiate Treants from
+        one another. Sometimes preferable to categories.
+
+        :Arguments:
+            *tags*
+                tags to add; must be single numbers, strings, or boolean
+                values; tags that are not these types are not added
+
+        """
+        # ensure tags are unique (we don't care about order)
+        # also they must be of a certain set of types
+        tags = set([tag for tag in tags
+                    if (isinstance(tag, (int, float, string_types, bool)) or
+                        tag is None)])
+
+        # remove tags already present in metadata from list
+        tags = tags.difference(set(self._record['tags']))
+
+        # add new tags
+        self._record['tags'].extend(tags)
+
+    @FileSerial._write
```

Comment:
I think the naming is a lot clearer now. This kind of grew out of the needs of the old PyTables backend, so the cleanup was certainly necessary.